cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -96,28 +96,29 @@ class GcsInputStream : public arrow::io::InputStream {
   // @name FileInterface
   Status Close() override {
     stream_.Close();
+    closed_ = true;
     return Status::OK();
   }
 
   Result<int64_t> Tell() const override {
-    if (!stream_) {
-      return Status::IOError("invalid stream");
-    }
+    if (closed()) return Status::Invalid("Cannot use Tell() on a closed stream");
     return stream_.tellg() + offset_;
   }
 
-  bool closed() const override { return !stream_.IsOpen(); }
+  bool closed() const override { return closed_ && !stream_.IsOpen(); }
```

Comment:
Is there a reason this is `&&` rather than `||`?

Reply:
Yes.  zero-length objects are "born closed" in `google-cloud-cpp`.  Probably a bug (or at least a very poorly thought API), but you can `read()` from them, while they  return `IsOpen() == false`.

Reply:
Ah, can you add a comment about that?

Reply:
Done. That got me thinking about the similar expression for output streams, and you were right in that case. Fixed that too.